core/tests/unittests/ray/test_resource_calculator.py

```diff
@@ -0,0 +1,145 @@
+    num_cpus = 32
+    num_jobs = 20
+    mem_available = psutil.virtual_memory().available
+    model_estimate_memory_usage = mem_available // 2.5  # allows 2 trials to run in parallel
+    
+    calculator = ResourceCalculatorFactory.get_resource_calculator(calculator_type='cpu')
+    assert type(calculator) == CpuResourceCalculator
+
+    resources_info = calculator.get_resources_per_job(
+        total_num_cpus=num_cpus,
+        num_jobs=num_jobs,
+        model_estimate_memory_usage=model_estimate_memory_usage,
+        minimum_cpu_per_trial=4,  # allows 8 jobs to run in parallel
+    )
+
+    expected_num_parallel_jobs = 2  # even user wants to run 20 jobs in prallel, cpu can run 8 jobs in parallel, memory only allows for 2 jobs
+    expected_resources_per_trial = dict(
+        cpu = 16,
+    )
+    expected_batches = 10
+
+    assert expected_resources_per_trial == resources_info['resources_per_job']
+    assert expected_num_parallel_jobs ==  resources_info['num_parallel_jobs']
+    assert expected_batches == resources_info['batches']
+    
+
+def test_gpu_calculator_no_bottleneck():
+    num_cpus = 32
+    num_gpus = 4
+    num_jobs = 20
+    
+    calculator = ResourceCalculatorFactory.get_resource_calculator(calculator_type='gpu')
+    assert type(calculator) == GpuResourceCalculator
+
+    resources_info = calculator.get_resources_per_job(
+        total_num_cpus=num_cpus,
+        total_num_gpus=num_gpus,
+        num_jobs=num_jobs,
+        minimum_cpu_per_trial=1, # allows 32 trials to run in parallel
+        minimum_gpu_per_trial=0.5,  # allows 8 trials to run in parallel
+    )
+
+    expected_num_parallel_jobs = 8
+    expected_resources_per_trial = dict(
+        cpu = 4,
+        gpu = 0.5,
+    )
+    expected_batches = 3
+
+    assert expected_resources_per_trial == resources_info['resources_per_job']
+    assert expected_num_parallel_jobs ==  resources_info['num_parallel_jobs']
+    assert expected_batches == resources_info['batches']
+    
+    
+def test_gpu_calculator_cpu_bottleneck():
+    num_cpus = 4
+    num_gpus = 4
+    num_jobs = 20
+
+    calculator = ResourceCalculatorFactory.get_resource_calculator(calculator_type='gpu')
+    assert type(calculator) == GpuResourceCalculator
+
+    resources_info = calculator.get_resources_per_job(
+        total_num_cpus=num_cpus,
+        total_num_gpus=num_gpus,
+        num_jobs=num_jobs,
+        minimum_cpu_per_trial=1, # allows 4 trials to run in parallel
+        minimum_gpu_per_trial=0.5,  # allows 8 trials to run in parallel
+    )
+
+    expected_num_parallel_jobs = 4
+    expected_resources_per_trial = dict(
+        cpu = 1,
+        gpu = 1,
+    )
+    expected_batches = 5
+
+    assert expected_resources_per_trial == resources_info['resources_per_job']
+    assert expected_num_parallel_jobs ==  resources_info['num_parallel_jobs']
+    assert expected_batches == resources_info['batches']
+
+
+def test_ray_lightning_gpu_calculator():
```

Comment:
Add tests checking assertion errors trigger when it is impossible to train with the given resources